### ISSUE TITLE
Intrepid2, Remove wasteful initialization

### ISSUE THREAD

**Issue by stanmoore1** [trilinos/Trilinos#6172]
## Enhancement
@trilinos/intrepid2

This type of code: https://github.com/trilinos/Trilinos/blob/master/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp#L251

```c++
      // Create ordinalToTag
      ordinalToTag = OrdinalTypeView2D("ordinalToTag", basisCard, tagSize);
```

wastes a CUDA kernel launch because Kokkos views are initialized to 0 by default, and then the view is set to -1 in the next line: https://github.com/trilinos/Trilinos/blob/master/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp#L254. 

```c++
      // Initialize with -1
      Kokkos::deep_copy( ordinalToTag, -1 );
```

These types of allocations should use `Kokkos::WithoutInitializing`

```c++
      ordinalToTag = OrdinalTypeView2D(Kokkos::view_alloc("ordinalToTag",Kokkos::WithoutInitializing), basisCard, tagSize);
```

**Comment by stanmoore1** [comment]
@bathmatt

**Comment by stanmoore1** [comment]
Likewise code like this: https://github.com/trilinos/Trilinos/blob/master/packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefModifyBasis.hpp#L291

should use `Kokkos::create_mirror_view_and_copy` instead of `Kokkos::create_mirror_view` + `Kokkos::deep_copy` to avoid unnecessary View initialization.

**Comment by jhux2** [comment]
@trilinos/intrepid2

**Comment by mperego** [comment]
OK, we'll work on that. Thanks for pointing it out.

**Comment by bathmatt** [comment]
@stanmoore1 is this the changes you made to improve EMPIRE initialization times??  Or is that different?

**Comment by stanmoore1** [comment]
@bathmatt no I have more changes to Intrepid2--I cached some device views in the basis class so they don't have to be allocated and copied every time.

**Comment by kyungjoo-kim** [comment]
@stanmoore1 @bathmatt 

Does the current basis construction hurt empire performance ? Tag information is created on host once when panzer basis functions are created. Unless the basis functions are high order (for hgrad the order is more than 2), the orientation modification does not invove. Even though the basis modifications with orientations are necessary when using high order elements,  the preparation (deep copy of tags to device) might be amortized as it changes all elements in a workset (assuming that the workset size is reasonably large). 

The tag information on the device certainly removes the copy overhead when the function is called repeatedly but I am not sure if this is the source of your performance problem. If it causes the performance problem, you may want to adjust the workset size as it implies that the code under-utilizes the GPU.

**Comment by stanmoore1** [comment]
This part of the code is getting called thousands/millions of times during EMPIRE initialization. This PR has a huge effect on EMPIRE initialization time:
```
~~~~~~OLD:~~~~~
|   Initialize: 454.845 - 87.2854% [1] {min=453.901, max=455.309, std dev=0.247654} <3, 1, 0, 0, 0, 4, 12, 42, 1, 1>

~~~~~~NEW:~~~~~
|   Initialize: 294.487 - 81.6049% [1] {min=294.326, max=294.646, std dev=0.0703561} <2, 6, 5, 3, 8, 22, 9, 3, 3, 3>
```

**Comment by stanmoore1** [comment]
By PR I mean: #6212.

**Comment by kyungjoo-kim** [comment]
I see the difference. It looks like that the performance hit might come from ``Intrepid2_LagrangianInterpolationDef.hpp`` assuming that the basis function does not require orientations. As you see in ``Intrepid2_OrientationToolsDefModifyBasis.hpp``, the code has a guard of ``if (basis->requireOrientation())`` while the interpolation function does not have such a guard. When the orientation is not required, the code would be a lot simpler. Also, the interpolation code includes both the setup and the batch-run for elements. I think that this would cause huge overhead too. @mperego You might want to consider to separate projection object that can be reused from the parallel for (apply phase) as well as code separation for basis functions that do not require orientations.

**Comment by stanmoore1** [comment]
Actually the basis function does require orientations so the hit is in `Intrepid2_OrientationToolsDefModifyBasis.hpp`.

**Comment by kyungjoo-kim** [comment]
Oh.... That is interesting. Then how come the function is called million times in the initialization phase ? Here when you mean by initialization, the empire usually calls the init function once and the init function invokes the orientation function million times ? Does the init mean that the init function is called per time step ? and the # of function calls is proportional to the # of time steps (millions) ? Do you know the workset size of the basis evaluations ?  I am wondering if the intrepid2 basis modification costs reasonable or if there is an interface issue in panzer/intrepid/empire.

**Comment by stanmoore1** [comment]
I didn't write that code so maybe @mbetten can comment. I just profiled it and saw thousands of calls to create the tags/ordinal views from `modifyBasisByOrientation()`. Each time the views are creates, a CUDA kernel is launched which has significant overhead. This is all during setup before doing any real timestepping.

**Comment by stanmoore1** [comment]
It is very possible that EMPIRE is doing something inefficient but I didn't write that code and am not familiar with Intrepid2 to know, someone else will have to answer.

**Comment by CamelliaDPG** [comment]
@stanmoore1, do you have the timings output from one of your profiling runs?  If you send it to me, I should be able to figure out where all those calls are coming from, and whether we can eliminate them.  I'll send you an email so that you have my address.

**Comment by rppawlo** [comment]
It's in both an empire function and the Panzer_ProjectToFaces. I didn't write the code either, but was asked to look into it. @CamelliaDPG - I can send you a link to the internal issues.

**Comment by kyungjoo-kim** [comment]
@rppawlo 

I am not sure about the empire case but I looked at the panzer project to face. It is written quite inefficient way when ``!use_fast_method_on_rectangular_hex_mesh``. If it uses the fast method, it does not call any intrepid2 function except for the quadrature point generations. If it does not use the fast method, the implementation cannot perform well. Some reference data and its temporal allocations should be done outside of the workset loop. This code does not include basis modification according to orientations. 

So, the basis modify function might be called from the empire.

**Comment by kyungjoo-kim** [comment]
I looked at the empire code. It looks like that the modifyBasisByOrientation is called from BlockEMSolverHelpers.cpp, line 564. This is not the right use case. The ots::modifyBasisByOrientation is supposed to handle all elements in a workset. In this code, the function is invoked for each element within the elemIter loop. This should not be used. The function launch cuda kernels for each element. That is why @stanmoore1 sees million times of deep copy. This is not intrepid issue but empire issue.

**Comment by stanmoore1** [comment]
You are still launching unnecessary CUDA kernels for initialization that don't need to happen. That is what I was trying to point out in this issue, it would be more efficient to use `Kokkos::WithoutInitializing`.

**Comment by CamelliaDPG** [comment]
This has been now been addressed by #6323.
